src/app/CASESessionManager.h

```diff
@@ -0,0 +1,120 @@
+#include <lib/core/CHIPCore.h>
+#include <lib/dnssd/DnssdCache.h>
+#include <lib/support/Pool.h>
+#include <messaging/ExchangeMgrDelegate.h>
+
+#include <lib/dnssd/Resolver.h>
+
+namespace chip {
+
+struct CASESessionManagerConfig
+{
+    DeviceProxyInitParams sessionInitParams;
+    Dnssd::DnssdCache<CHIP_CONFIG_MDNS_CACHE_SIZE> * dnsCache = nullptr;
+};
+
+/**
+ * This class provides the following
+ * 1. Manage a pool of operational device proxy objects for peer nodes that have active message exchange with the local node.
+ * 2. The pool contains atmost one device proxy object for a given peer node.
+ * 3. API to lookup an existing proxy object, or allocate a new one by triggering session establishment with the peer node.
+ * 4. During session establishment, trigger node ID resolution (if needed), and update the DNS-SD cache (if resolution is
+ * successful)
+ */
+class CASESessionManager : public Messaging::ExchangeMgrDelegate, public Dnssd::ResolverDelegate
+{
+public:
+    CASESessionManager(CASESessionManagerConfig & params)
+    {
+        VerifyOrReturn(params.sessionInitParams.Validate() == CHIP_NO_ERROR);
+
+        mConfig      = params;
+        mInitialized = true;
+    }
+
+    virtual ~CASESessionManager() {}
+
+    void Shutdown() { mInitialized = false; }
+
+    /**
+     * Find an existing session for the given node ID, or trigger a new session request.
+     * The caller can optionally provide `onConnection` and `onFailure` callback objects. If provided,
+     * these will be used to inform the caller about successful or failed connection establishment.
+     * If the connection is already established, the `onConnection` callback will be immediately called.
+     */
+    CHIP_ERROR FindOrEstablishSession(NodeId nodeId, Callback::Callback<OnDeviceConnected> * onConnection,
+                                      Callback::Callback<OnDeviceConnectionFailure> * onFailure);
+
+    OperationalDeviceProxy * FindExistingSession(NodeId nodeId);
+
+    void ReleaseSession(NodeId nodeId);
+
+    FabricInfo * GetFabricInfo()
+    {
+        if (!mInitialized)
+        {
+            return nullptr;
+        }
+
+        return mConfig.sessionInitParams.fabricInfo;
+    }
+
+    /**
+     * This API triggers the DNS-SD resolution for the given node ID. The node ID will be looked up
+     * on the fabric that was configured for the CASESessionManager object.
+     *
+     * The results of the DNS-SD resolution request is provided to the class via `ResolverDelegate`
+     * implementation of CASESessionManager.
+     */
+    CHIP_ERROR ResolveDeviceAddress(NodeId nodeId);
+
+    /**
+     * This API returns the address and port for the given node ID. It is expected that there is
+     * an ongoing session with the peer node. If the session doesn't exist, the API will return
+     * `CHIP_ERROR_NOT_CONNECTED` error.
+     */
+    CHIP_ERROR GetDeviceAddressAndPort(NodeId nodeId, Inet::IPAddress & addr, uint16_t & port);
```

Comment:
Using PeerAddress as output parameter would give us an interface that can provide the address without ambiguity.  As this sits, this does not provide protocol (UDP or TCP) or interface ID, which is necessary for link local addresses.